Spigot-Server-Patches/0615-Optimized-villager-pathfinding.patch

```diff
@@ -0,0 +1,282 @@
+"villagerBehaviorMaxCooldown". The cooldown will be a random number of ticks between these two values. The default config values do not change the vanilla behavior at all.
+
+diff --git a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+index ae5ed3bd0b663092a4658b24cbd69d37b4e141cb..61fb44c2efa2771e05c9c0d4a833385194316a47 100644
+--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
++++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+@@ -709,4 +709,14 @@ public class PaperWorldConfig {
+     private void fixCuringExploit() {
+         fixCuringZombieVillagerDiscountExploit = getBoolean("game-mechanics.fix-curing-zombie-villager-discount-exploit", fixCuringZombieVillagerDiscountExploit);
+     }
++
++    public int villageSearchRadius = 48;
++    public int villagerBehaviorMinCooldown = 0;
++    public int villagerBehaviorMaxCooldown = 20;
++    private void villagerBehaviorSettings() {
++        villageSearchRadius = getInt("villager-behavior.village-search-radius", villageSearchRadius);
++        villagerBehaviorMinCooldown = getInt("villager-behavior.cooldown-min", villagerBehaviorMinCooldown);
++        villagerBehaviorMaxCooldown = getInt("villager-behavior.cooldown-max", villagerBehaviorMaxCooldown);
++    }
++
+ }
+diff --git a/src/main/java/net/minecraft/server/BehaviorFindPosition.java b/src/main/java/net/minecraft/server/BehaviorFindPosition.java
+index 63a761ebef80d4af09cdc2682e496d78492c4a3a..ce836bfd8c56c7ff1e5eb53fb487b064da2a8a0d 100644
+--- a/src/main/java/net/minecraft/server/BehaviorFindPosition.java
++++ b/src/main/java/net/minecraft/server/BehaviorFindPosition.java
+@@ -48,7 +48,7 @@ public class BehaviorFindPosition extends Behavior<EntityCreature> {
+         if (this.d && entitycreature.isBaby()) {
+             return false;
+         } else if (this.f == 0L) {
+-            this.f = entitycreature.world.getTime() + (long) worldserver.random.nextInt(20);
++            this.f = entitycreature.world.getTime() + worldserver.paperConfig.villagerBehaviorMinCooldown + (long)worldserver.random.nextInt(worldserver.paperConfig.villagerBehaviorMaxCooldown - worldserver.paperConfig.villagerBehaviorMinCooldown); // Paper - Option to slow down villagers
+             return false;
+         } else {
+             return worldserver.getTime() >= this.f;
+@@ -56,7 +56,7 @@ public class BehaviorFindPosition extends Behavior<EntityCreature> {
+     }
+ 
+     protected void a(WorldServer worldserver, EntityCreature entitycreature, long i) {
+-        this.f = i + 20L + (long) worldserver.getRandom().nextInt(20);
++        this.f = i + 20L + (long)worldserver.random.nextInt(worldserver.paperConfig.villagerBehaviorMaxCooldown - worldserver.paperConfig.villagerBehaviorMinCooldown); // Paper - Option to slow down villagers
```

Comment:
I am not fully awake yet, but shouldn't this be `i + 20 + min + random(max-min)`? (the + min is missing)

Reply:
You're totally right, I've fixed this now

Reply:
now you removed the +20, I think that still needs to be there.

vanilla is i + 20 + random between 0 and 20
we want i + 20 + random between min and max, which translates into i + 20 + min + random(max-min)